src/wallet/walletdb.cpp

```diff
@@ -842,6 +747,269 @@ static DBErrors LoadWalletFlags(CWallet* pwallet, DatabaseBatch& batch) EXCLUSIV
+    CDataStream ssValue(SER_DISK, CLIENT_VERSION);
+
+    CDataStream prefix(0, 0);
+    prefix << key;
+    std::unique_ptr<DatabaseCursor> cursor = batch.GetNewPrefixCursor(prefix);
+    if (!cursor) {
+        pwallet->WalletLogPrintf("Error getting database cursor for '%s' records\n", key);
+        result.m_result = DBErrors::CORRUPT;
+        return result;
+    }
+
+    while (true) {
+        DatabaseCursor::Status status = cursor->Next(ssKey, ssValue);
+        if (status == DatabaseCursor::Status::DONE) {
+            break;
+        } else if (status == DatabaseCursor::Status::FAIL) {
+            pwallet->WalletLogPrintf("Error reading next '%s' record for wallet database\n", key);
+            result.m_result = DBErrors::CORRUPT;
+            return result;
+        }
+        std::string type;
+        ssKey >> type;
+        assert(type == key);
+        if ((result.m_result = std::max(result.m_result, load_func(pwallet, ssKey, ssValue, result.m_error))) == DBErrors::CORRUPT) {
+            pwallet->WalletLogPrintf("%s\n", result.m_error);
+        }
+        ++result.m_records;
+    }
+    return result;
+}
+
+static DBErrors LoadLegacyWalletRecords(CWallet* pwallet, DatabaseBatch& batch, int last_client) EXCLUSIVE_LOCKS_REQUIRED(pwallet->cs_wallet)
+{
+    AssertLockHeld(pwallet->cs_wallet);
+    DBErrors result = DBErrors::LOAD_OK;
+
+    // Make sure descriptor wallets don't have any legacy records
+    if (pwallet->IsWalletFlagSet(WALLET_FLAG_DESCRIPTORS)) {
+        for (const auto& type : DBKeys::LEGACY_TYPES) {
+            CDataStream key(SER_DISK, CLIENT_VERSION);
+            CDataStream value(SER_DISK, CLIENT_VERSION);
+
+            CDataStream prefix(0, 0);
+            prefix << type;
+            std::unique_ptr<DatabaseCursor> cursor = batch.GetNewPrefixCursor(prefix);
+            if (!cursor) {
+                pwallet->WalletLogPrintf("Error getting database cursor for '%s' records\n", type);
+                return DBErrors::CORRUPT;
+            }
+
+            DatabaseCursor::Status status = cursor->Next(value, value);
+            if (status != DatabaseCursor::Status::DONE) {
+                pwallet->WalletLogPrintf("Error: Unexpected legacy entry found in descriptor wallet %s. The wallet might have been tampered with or created with malicious intent.\n", pwallet->GetName());
+                return DBErrors::UNEXPECTED_LEGACY_ENTRY;
+            }
+        }
+
+        return DBErrors::LOAD_OK;
+    }
+
+    // Load HD Chain
+    CHDChain chain;
+    if (batch.Read(DBKeys::HDCHAIN, chain)) {
+        pwallet->GetOrCreateLegacyScriptPubKeyMan()->LoadHDChain(chain);
+    }
+
+    // Load unencrypted keys
+    LoadResult key_res = LoadRecords(pwallet, batch, DBKeys::KEY,
+        [] (CWallet* pwallet, CDataStream& key, CDataStream& value, std::string& err) {
+        if (!LoadKey(pwallet, key, value, err)) {
+            return DBErrors::CORRUPT;
+        }
+        return DBErrors::LOAD_OK;
```

Comment:
In 4f83673a:

tiny styling nit (no need to do it if you don't re-touch):
```c++
return LoadKey(pwallet, key, value, err) ? DBErrors::LOAD_OK : DBErrors::CORRUPT;
```

(same for the encrypted keys load)

Reply:
Done